include/vast/Conversion/Common/Passes.hpp

```diff
@@ -89,4 +91,79 @@ namespace vast {
         // interface with pass base
         void runOnOperation() override { run_on_operation(); }
     };
+
+    // Sibling of the above module for passes that go to the LLVM dialect.
+    // Example usage:
+    //
+    // struct ExamplePass : ModuleLLVMConversionPassMixin< ExamplePass, ExamplePassBase > {
+    //     using base = ModuleLLVMConversionPassMixin< ExamplePass, ExamplePassBase >;
+    //
+    //     static conversion_target create_conversion_target(MContext &context) {
+    //         conversion_target target(context);
+    //         // setup target here
+    //         return target;
+    //     }
+    //
+    //     static void populate_conversions(rewrite_pattern_set &patterns) {
+    //         base::populate_conversions_base<
+    //             // pass conversion type_lists here
+    //         >(patterns);
+    //     }
+    //     static void set_llvm_options(mlir::LowerToLLVMOptions &llvm_options) {
+    //          llvm_options.myOption = my_value;
+    //     }
+    // }
+    //
+    template< typename derived_t, template< typename > typename base_t >
+    struct ModuleLLVMConversionPassMixin : base_t< derived_t > {
+
+        using base = base_t< derived_t >;
+
+        using base::getContext;
+        using base::getOperation;
+        using base::signalPassFailure;
+
+        using rewrite_pattern_set = mlir::RewritePatternSet;
+
+        using type_converter = util::tc::LLVMTypeConverter;
+
+        template< typename list >
+        static void populate_conversions_impl(
+            rewrite_pattern_set &patterns, type_converter &converter)
+        {
+            if constexpr ( list::empty ) {
+                return;
+            } else {
+                patterns.add< typename list::head >(converter);
+                return populate_conversions_base< typename list::tail >(patterns, converter);
+            }
+        }
+
+        template< typename ...lists  >
+        static void populate_conversions_base(rewrite_pattern_set &patterns, type_converter &converter) {
```

Comment:
Make whatever method needs to be overwritten `virtual`.

Reply:
Right now no methods need to be overwritten. In 6915f1d901be8c4182ba13e26d7d95f7ff572415 I modified it so that the method implemented by the user is `populate_conversions` and the method in the base class is `populate_conversions_base` (so that it explodes if there is no `populate_conversions` method). Should I rename it back?

Reply:
Sorry, I missed that! This is fine.